subsys/mgmt/mcumgr/grp/os_mgmt/src/os_mgmt.c

```diff
@@ -310,7 +310,7 @@ static int os_mgmt_taskstat_read(struct smp_streamer *ctxt)
  */
 static void os_mgmt_reset_work_handler(struct k_work *work)
 {
-	sys_reboot(SYS_REBOOT_WARM);
+	sys_reboot(SYS_REBOOT_COLD);
```

Comment:
@nordicjm Do we specify type for reboot?

Reply:
Note: the meaning of cold/warm reboot was not clear (there were no docs at all), so it was used randomly. The meaning of "warm" reboot is still not much clearer (it's probably driven by PSCI warm reboot).

Reply:
> This function is/was a bit of a lie, WARM/COLD is not what this function has done, warm reboot _should_ be jumping to the restart vector of the application, cold reboot _should_ be using the CMSIS restart function, but actually what this does is CMSIS restart in both cases but (depending on a Kconfig) sets one of the GPREGRET registers to a value. Nothing in-tree uses it and neither does nor has MCUboot ever used it, it should never have been added and was not fit for purpose.

Note that this is your interpretation of warm/cold because it was not documented at all.

Reply:
warm/cold reboot are standard terms, you would expect a cold reboot to reset the hardware e.g. watchdog timer, you would not expect a warm reboot to do that

Reply:
cold reboot is basically what you expect to happen on powering up a device; warm reset may be just jumping at specified vector that will reinitialize the software you are warm rebooting, which means that all set-once-per-runtime things, like some hardware assisted trusted enclaves and so on, may persist in state they have been set by powerup.

Reply:
one of those things defined in simple terms on wikipedia: https://en.wikipedia.org/wiki/Reboot. We end up having this discussion everytime the API is touched. This API has been with us since Zephyr was launched and if memory serves right, it was mostly used on x86 systems, but then got adopted by various platforms. Warm and Cold are probably very limited in covering the various use-cases we have these days.

Maybe it time to deprecate this and move to something a bit more suitable to all platforms and usecases we have now.

Reply:
According to wikipedia description, we’ve never had cold reboot then, because it implies “power off” as well. The main problem is that Zephyr never defined what its reboot types mean, and as a result, implementations differ and there’s a mix of warm/cold usage because nobody knows what to choose.

Reply:
hmm, this what we had in Zephyr v1.0.0:

```
static inline void cold_reboot(void)
{
        uint8_t reset_value = SYS_X86_RST_CNT_CPU_RST | SYS_X86_RST_CNT_SYS_RST |
                                                        SYS_X86_RST_CNT_FULL_RST;
        sys_out8(reset_value, SYS_X86_RST_CNT_REG);
}

void sys_arch_reboot(int type)
{
        switch (type) {
        case SYS_REBOOT_COLD:
                cold_reboot();
                break;
        default:
                /* do nothing */
                break;
        }
}
```

So, never is not accurate, unless this code does not do what was intended.

Reply:
> According to wikipedia description, we’ve never had cold reboot then, because it implies “power off” as well. The main problem is that Zephyr never defined what its reboot types mean, and as a result, implementations differ and there’s a mix of warm/cold usage because nobody knows what to choose.

It does not have to be power off. Cold reset on old AT and PS/2 platforms was controlled by external 8042 chip that could be either triggered by software or short-circuited by reset pin on the device enclosure.

Many MCUs have external MCLR/RESET pin that can be controlled by external device, and works similarly to power cycle, as it is often intended to be controlled by specialized Reset IC. I have used a reset circuit where MCU, not running Zephyr (sorry),  would trigger external reset circuit doing t1-wait->t2-hold->release for itself and other MCUs with all MCLR connected to the same line, basically power cycling itself and others.

The MCUs get more and more complicated and have internal DC/DC converters and power rails that can be disabled and enabled by PM controller for various reasons; if there is an MCU with PM controller where you can trigger it to power down all rails for a moment that would be a cold reset, from the software perspective.

Shouldn't we have PM  experts from IC vendors, contributing to Zephyr, gathered in single room and have them discuss what they have and how can we address it?

Reply:
> Shouldn't we have PM experts from IC vendors, contributing to Zephyr, gathered in single room and have them discuss what they have and how can we address it?

That would be ideal, but... do we have access to such expertise?

Reply:
The thing is, with the kind of µCs Zephyr supports in general, there is a lot of possible "reset" options.
Starting with: Are you resetting just the MCU which is executing that code? (maybe with its closest linked peripherals, maybe its whole "domain"?) the whole SOC? more ICs than just the SOC?
Then even if you may be resetting "the whole SOC", that SOC may easily support different reset modes (will it reset GPIOs configuration, security enclaves,..), and the SOC may not allow all resets to be triggered from any states or from any of its µCs.
So I would lean to leave this quite open and overridable by the SOC/board. Even if at some point some kind of resets being so common as to be explicitly listed at the kernel level, I'd still let SOCs and boards expand that list.

Reply:
The problem is that if we open `sys_reboot()` to vendor-specific reset options, the API becomes even less portable. Refer to https://github.com/zephyrproject-rtos/zephyr/pull/60920/commits/a7e841ba31842295dfcfcb76503f0b02557f13d1. Remember this API is called from generic code that has no specific system knowledge.

Reply:
> The problem is that if we open sys_reboot() to vendor-specific reset options, the API becomes even less portable.

Well.. not necessarily. At Zephyr's level we'd define the basic reset modes any code can call.
If a ARCH or SOC family supports some extra reset modes, it would only be then code from that ARCH/SOC family (or its drivers) or apps which specifically target those who would call with that reset type.
We also specify that a not supported reset mode is received, we fall back to another type of reset.